components/hooks.ts

```diff
@@ -0,0 +1,320 @@
+import { ChainRegistryClient } from '@penumbra-labs/registry';
+import {
+  PenumbraClient,
+  type PenumbraManifest,
+  PenumbraRequestFailure,
+  PenumbraState,
+} from '@penumbra-zone/client';
+import { TendermintProxyService, ViewService } from '@penumbra-zone/protobuf';
+import {
+  type AssetId,
+  type Metadata,
+  ValueView,
+} from '@penumbra-zone/protobuf/penumbra/core/asset/v1/asset_pb';
+import type { TransactionView } from '@penumbra-zone/protobuf/penumbra/core/transaction/v1/transaction_pb';
+import { useQuery } from '@tanstack/react-query';
+import { uniqBy } from 'es-toolkit';
+import { useEffect, useState } from 'react';
+import { client } from './penumbra';
+import { useQuestStore } from './store';
+
+export const useWalletManifests = () => {
+  return useQuery<Record<string, PenumbraManifest>>({
+    queryKey: ['providerManifests'],
+    staleTime: 0,
+    queryFn: async () => {
+      const res = PenumbraClient.getProviderManifests();
+      const resolvedManifests = await Promise.all(
+        Object.entries(res).map(async ([key, promise]) => {
+          const value = await promise;
+          return [key, value];
+        }),
+      );
+      return Object.fromEntries(resolvedManifests ?? {});
+    },
+  });
+};
+
+export const useConnect = () => {
+  const {
+    data: connected,
+    isLoading: connectionLoading,
+    refetch,
+  } = useQuery({
+    queryKey: ['connection'],
+    queryFn: async () => {
+      const providers = PenumbraClient.getProviders();
+      const connected = Object.keys(providers).find((origin) =>
+        PenumbraClient.isProviderConnected(origin),
+      );
+      if (!connected) return undefined;
+      try {
+        await client.connect(connected);
+        return connected;
+      } catch (error) {
+        return undefined;
+      }
+    },
+  });
+
+  const onConnect = async (origin: string) => {
+    try {
+      await client.connect(origin);
+      await refetch();
+    } catch (error) {
+      if (error instanceof Error && error.cause) {
+        if (error.cause === PenumbraRequestFailure.Denied) {
+          alert(
+            'Connection denied: you may need to un-ignore this site in your extension settings.',
```

Comment:
suggestion: could we also add a note about refreshing the screen?

https://github.com/user-attachments/assets/1d63b18c-ae8f-48a9-9a12-ba7db411a7e1